packages/amplify_storage_s3/lib/method_channel_storage_s3.dart

```diff
@@ -102,39 +100,80 @@ class AmplifyStorageS3MethodChannel extends AmplifyStorageS3 {
       return result;
     } on PlatformException catch (e) {
-      //TODO: Convert to StorageError and throw.
-      throw (e);
+      throw (_convertToStorageException(e));
     }
   }
 
   UploadFileResult _formatUploadFileResult(Map<String, dynamic> result) {
-    return UploadFileResult(key: result["key"]);
+    if (result['key'] != null) {
+      return UploadFileResult(key: result["key"]);
+    } else {
+      throw (_prepareExceptionForMalformedResult(
+          methodName: 'UploadFile', fieldName: 'key'));
+    }
   }
 
   GetUrlResult _formatGetUrlResult(Map<String, dynamic> result) {
-    return GetUrlResult(url: result["url"]);
+    if (result['url'] != null) {
+      return GetUrlResult(url: result['url']);
+    } else {
+      throw (_prepareExceptionForMalformedResult(
+          methodName: 'GetUrl', fieldName: 'url'));
+    }
   }
 
   RemoveResult _formatRemoveResult(Map<String, dynamic> result) {
-    return RemoveResult(key: result["key"]);
+    if (result['key'] != null) {
+      return RemoveResult(key: result['key']);
+    } else {
+      throw (_prepareExceptionForMalformedResult(
+          methodName: 'Remove', fieldName: 'key'));
+    }
   }
 
-  ListResult _formatListResult(Map<String, dynamic> response) {
+  ListResult _formatListResult(Map<String, dynamic> result) {
     List<StorageItem> items = [];
-    var mapItems = response['items'];
+    if (result['items'] == null) {
+      throw (_prepareExceptionForMalformedResult(
+          methodName: 'List', fieldName: 'items'));
+    }
+    var mapItems = result['items'];
     for (Map item in mapItems) {
-      StorageItem storageItem = StorageItem(
-          key: item["key"],
-          eTag: item["eTag"],
-          lastModified: DateTime.parse(item["lastModified"]),
-          size: item["size"]);
-      items.add(storageItem);
+      if (item["key"] != null) {
+        StorageItem storageItem = StorageItem(
+            key: item["key"],
+            eTag: item["eTag"],
+            lastModified: DateTime.parse(item["lastModified"]),
+            size: item["size"]);
+        items.add(storageItem);
+      } else {
+        throw (_prepareExceptionForMalformedResult(
+            methodName: 'List', fieldName: 'key'));
+      }
     }
-    ListResult result = ListResult(items: items);
-    return result;
+    return ListResult(items: items);
   }
 
   DownloadFileResult _formatDownloadFileResult(Map<String, dynamic> result) {
-    return DownloadFileResult(file: File(result["path"]));
+    if (result['path'] != null) {
+      return DownloadFileResult(file: File(result['path']));
+    } else {
+      throw (_prepareExceptionForMalformedResult(
+          methodName: 'DownloadFile', fieldName: 'path'));
+    }
+  }
+
+  StorageException _prepareExceptionForMalformedResult(
+      {@required String methodName, @required String fieldName}) {
+    final Map errorDetails = <String, String>{
+      '$methodName operation failed': '$fieldName cannot be null'
+    };
```

Comment:
It's difficult to understand this errorMap structure where both key and values are error details.

Suggested reformatting
```
{ 
   'operation_failed': '$methodName',
   'error_field': '$fieldName' 
}
```

Also helpful to know what PlatformException::details would be like to get consistent StorageException

Reply:
Updated as per our offline discussion.